online-inference/gpt-j-deepspeed/service/service.py

```diff
@@ -0,0 +1,292 @@
+            replace_method="auto",
+            replace_with_kernel_inject=True,
+            enable_cuda_graph=True,
+        )
+        logger.info(f"DeepSpeed initialized.")
+
+        logger.info(
+            f"Model is ready in {str(time.perf_counter() - start_time)} seconds."
+        )
+        self.ready = True
+
+    def configure_request(self, request):
+        request_params = model_params.copy()
+
+        if "parameters" in request:
+            parameters = request["parameters"]
+            for k, pv in parameters.items():
+                pk = k.upper()
+                if pk in request_params:
+                    logger.debug(
+                        f"Parameter {pk} changed from {request_params[pk]} to {pv}"
+                    )
+                    request_params[pk] = pv
+
+        if "benchmark" in request:
+            request_params["BENCHMARK"] = True
+            request_params["MIN_LENGTH"] = request_params["MAX_LENGTH"]
+        else:
+            request_params["BENCHMARK"] = False
+
+        if "instances" in request:
+            request_params["INSTANCES"] = request["instances"]
+
+        return request_params
+
+    def _predict(self, request_params):
+        return {
+            "predictions": self.generator(
+                request_params["INSTANCES"][0],
+                do_sample=False,
+                min_length=request_params["MIN_LENGTH"],
+                max_length=request_params["MAX_LENGTH"],
+                temperature=request_params["TEMPERATURE"],
+                top_k=request_params["TOP_K"],
+                top_p=request_params["TOP_P"],
+                repetition_penalty=request_params["REPETITION_PENALTY"],
+                pad_token_id=self.tokenizer.eos_token_id,
+            )
+        }
+
+    def predict(self, request: Dict) -> Dict:
+        request_params = self.configure_request(request)
+        if request_params["BENCHMARK"] is True:
+            return self.benchmark(request_params)
+        else:
+            return self._predict(request_params)
+
+    def load_dataset(self):
+        data = open("the-time-machine.txt", "r").read()
+        tokens = self.tokenizer.encode(data)
+        sequence_start = random.randrange(len(tokens) - 2048)
+        sequence_end = sequence_start + 2048
+        return tokens[sequence_start:sequence_end]
+
+    def benchmark(self, request_params):
+        if self.dataset is None:
+            self.dataset = self.load_dataset()
+
+        assert (
+            request_params["BENCHMARK_SEQUENCE_LENGTH"] < request_params["MAX_LENGTH"]
+        )
+
+        sequence_start = random.randrange(
+            len(self.dataset) - request_params["BENCHMARK_SEQUENCE_LENGTH"]
+        )
+        sequence_end = sequence_start + request_params["BENCHMARK_SEQUENCE_LENGTH"]
+        random_sequence_encoded = self.dataset[sequence_start:sequence_end]
+        random_sequence = self.tokenizer.decode(random_sequence_encoded)
```

Comment:
Can you explain how you are randomizing the sequences here? Does `randrange` require two args: https://www.w3schools.com/python/ref_random_randrange.asp

Reply:
`randrange()` can take 1-3 args. 1 arg just returns a random value between 0 and your arg N. 

We pick a random start that is at most within `BENCHMARK_SEQUENCE_LENGTH` of the end of the `self.dataset`, so that when we add `BENCHMARK_SEQUENCE_LENGTH` to the start, we don't go out of range.

Then we grab the slice of `self.dataset` from `start` to `end` indexes. 

The 3 arg form of `randrange()` could also be used, but being limited to incrementing by the `step` arg would limit the number of random sequences we could get.